Expands Tests, Adds Exceptions, Remove Parquet and Square Matrix Support

emarinier opened the pull request:
Adds various tests, expanded input and error handling with exceptions, and removed parquet and square matrix support (https://github.com/phac-nml/genomic_address_service/commit/a368af631b44fdd0e08524da31db5a65580564fc).

**Notable Changes**
- Thresholds must be strictly decreasing.
- Exceptions and handling instead of `sys.exit()`.
- Minor changes to main/argparse/args/run handling.
- Static class variables have been changed to dynamic instance variables.
- Renamed many instances of `delimeter` to `delimiter`.
- Previously, delimiters provided on the command line to `gas call` were not being used. This has been fixed.
- Expanded `gas call` delimiter` error handling.
- Removed parquet references and support.
- Remove square distance matrix input references and support.

**test_mcluster.py**
- basic
- single linkage wikipedia example
- thresholds that lead to the same clustering
- non-decreasing thresholds throws exception
- invalid and missing threshold throws exception
- various delimiters
- missing and/or empty distance matrix throws exception
- invalid linkage method throws exception
- 14 thresholds works as expected
- single linkage
- complete linkage
- average linkage

**test_call.py**
- basic
- multiple thresholds leading to same clusters
- non-decreasing thresholds throws exception
- invalid and missing threshold throws exception
- various delimiters
- delimiter mismatch throws exception
- delimiter too long throws exception
- invalid delimiters throw exception
- error handling address errors
- various missing inputs throw exceptions
- invalid linkage throws exception
- bad file extensions throw exceptions
- good extensions
- small data set

apetkau commented on tests/test_call.py:
```diff
@@ -22,13 +29,1242 @@ def test_run_call(tmp_path):
+        assert thresholds_json["1"] == 3.0
+        assert thresholds_json["2"] == 0.0
+
+def test_method_average(tmp_path):
```
Thanks so much for adding these tests. This looks great.

For `average` and `complete` linkage test cases, do you think you could add cases where there is a difference in assigned addresses? That is where `average` or `complete` give a different answer from `single`?

emarinier replied:
Yes, I can, I'll see if the Wikipedia dataset gives different answers, otherwise it might take a bit to craft and manually check what the answer should be.

emarinier replied:
**Single**

```
gas mcluster -t 45,40,35,30,25,20,15,10,5,0 -i tests/data/matrix/wikipedia-single.tsv -o gas_out --force --method single
cp gas_out/clusters.text tests/data/clusters/wikipedia-single.tsv
```

```
id	address	level_1	level_2	level_3	level_4	level_5	level_6	level_7	level_8	level_9	level_10
a	1.1.1.1.1.1.1.1.1.1	1	1	1	1	1	1	1	1	1	1
b	1.1.1.1.1.1.2.2.2.2	1	1	1	1	1	1	2	2	2	2
c	1.1.1.1.1.2.3.3.3.3	1	1	1	1	1	2	3	3	3	3
d	1.1.1.1.2.4.5.5.5.5	1	1	1	1	2	4	5	5	5	5
e	1.1.1.1.1.3.4.4.4.4	1	1	1	1	1	3	4	4	4	4
```

**Average**

```
gas mcluster -t 45,40,35,30,25,20,15,10,5,0 -i tests/data/matrix/wikipedia-single.tsv -o gas_out --force --method average
cp gas_out/clusters.text tests/data/clusters/wikipedia-average.tsv
```

```
id	address	level_1	level_2	level_3	level_4	level_5	level_6	level_7	level_8	level_9	level_10
a	1.1.1.1.1.1.1.1.1.1	1	1	1	1	1	1	1	1	1	1
b	1.1.1.1.1.1.2.2.2.2	1	1	1	1	1	1	2	2	2	2
c	1.1.1.2.2.3.4.4.4.4	1	1	1	2	2	3	4	4	4	4
d	1.1.1.2.3.4.5.5.5.5	1	1	1	2	3	4	5	5	5	5
e	1.1.1.1.1.2.3.3.3.3	1	1	1	1	1	2	3	3	3	3
```

**Complete**

```
gas mcluster -t 45,40,35,30,25,20,15,10,5,0 -i tests/data/matrix/wikipedia-single.tsv -o gas_out --force --method complete
cp gas_out/clusters.text tests/data/clusters/wikipedia-complete.tsv
```

```
id	address	level_1	level_2	level_3	level_4	level_5	level_6	level_7	level_8	level_9	level_10
a	1.1.1.1.1.1.1.1.1.1	1	1	1	1	1	1	1	1	1	1
b	1.1.1.1.1.1.2.2.2.2	1	1	1	1	1	1	2	2	2	2
c	1.2.2.2.2.3.4.4.4.4	1	2	2	2	2	3	4	4	4	4
d	1.2.2.2.3.4.5.5.5.5	1	2	2	2	3	4	5	5	5	5
e	1.1.1.1.1.2.3.3.3.3	1	1	1	1	1	2	3	3	3	3
```

**Remarks**

The original distance matrix for `mcluster` is "wikipedia-single.tsv", but it's just a distance matrix that was used for a single clustering example. It can be used for other linkage distance methods.

Since the cluster address for `d` is different for each linkage type, I will introduce a new ID `f` that matches `d` in distances to other IDs, so that it is assigned the same cluster address (which will be different for each linkage).

single: `1.1.1.1.2.4.5.5.5.5`
average: `1.1.1.2.3.4.5.5.5.5`
complete: `1.2.2.2.3.4.5.5.5.5`

**Pairwise Distances File(s)**

```
query_id	ref_id	dist
f	a	31
f	b	34
f	c	28
f	d	0
f	e	43
f	f	0
```

**Updated**: 765fae8

apetkau commented on README.md:
```diff
@@ -141,43 +138,22 @@ There are a number of arguments that are specific for each command. They can be
 
-## Data Input/formats
-
-### Square distance matrix
```
Square distance matrices are still used as part of `gas mcluster`, so I think we should still have a description of this format in the docs.

Was there no description of the three-column input type? Maybe we should have that included here as well (used for `gas call`).

emarinier replied:
Oh that's right. I'll look over this again and re-add in a more clear way.

emarinier replied:
I don't believe there was a description for the 3-column input, but maybe I'm misremembering.

apetkau replied:
Thanks so much. I agree. Expanding can be in another PR.

apetkau reviewed (changes requested):
Thanks so much @emarinier for your amazing work on this 😄 . It's very much appreciated.

In addition to in-line comments, could you also:
1. Update version
2. Add entry to CHANGELOG.md

emarinier commented on tests/test_reader.py:
```diff
@@ -95,16 +95,6 @@ def test_dist_reader_functionality_pq(query_table_pq):
     assert list(d.read_data()) == [{'sampleQ': {'sample1': 1.0, 'sample2': 1.0, 'sample3': 2.0}, 'sampleN': {'sample1': 1.0, 'sample2': 1.0, 'sample3': 2.0}}]
 
-
-def test_guess_file_type(test_class, query_table_tsv):
```
Yeah, what's basically happened was there was a lot of vestigial code that supported different input file types/formats, but they were not implemented enough to function. I removed the vestigial code, considering the only supported into type is text/3-column pairwise distances.

emarinier commented on genomic_address_service/call.py:
```diff
@@ -29,68 +29,71 @@ class CustomFormatter(ArgumentDefaultsHelpFormatter, RawDescriptionHelpFormatter
     parser.add_argument('-o','--outdir', type=str, required=True, help='Output directory to put cluster results')
-    parser.add_argument('-u', '--outfmt', type=str, required=False, help='Output format for assignments [text, parquet]',default='text')
-    parser.add_argument('-l', '--delimeter', type=str, required=False, help='delimeter desired for nomenclature code',default=".")
+    parser.add_argument('-l', '--delimiter', type=str, required=False, help='The delimiter used within addresses in the input cluster file, as well as the delimiter to use for addresses in the output.', default=".")
```
Added in 44a55fb.

emarinier commented on genomic_address_service/call.py:
```diff
@@ -104,16 +107,21 @@ def run_call(config):
     cluster_assignments = obj.memberships_dict
 
-    run_data['result_file'] = os.path.join(outdir, "results.{}".format(outfmt))
+    run_data['result_file'] = os.path.join(outdir, "results.text")
```
#40

kbessonov1984 replied:
I agree that TSV is a better extension than tsv

apetkau replied:
Thanks Eric.

emarinier commented on genomic_address_service/call.py:
```diff
@@ -122,7 +130,13 @@ def run_call(config):
+        call(vars(cmd_args))
+
+    except Exception as exception:
+        print("Exception: " + str(exception))
```
The issue with the previous implementation of printing errors to standard out and immediately `sys.exit()`ing is that it makes for uncomfortable design patterns when calling the software from tests (or any future uses that would call `call` directly). The calling function doesn't have a choice in the entire Python execution ending and retrieving error messages for testing purposes is painful.

The choice to catch the exceptions, print, and `sys.exit()` in this run command, was to try to more closely preserve the original behaviour: `sys.exit()` without an exception trace stack printed to the user.

I don't mind removing the except handling here, or changing it to `sys.exit(1)`, but it will change the behaviour slightly.

emarinier replied:
Updated in afdc617

kbessonov1984 replied:
Here we area interested to print Exceptions without stopping a program to standard output? We do not raise Exceptions?

emarinier replied:
Can you clarify? The very next line of code is `sys.exit(1)`:

https://github.com/phac-nml/genomic_address_service/blob/ed6e5154eff43fcc807cf53c15cc0a2ee104ae39/genomic_address_service/call.py#L134-L142

emarinier replied:
I'll also note that the change from `sys.exit()` to `sys.exit(1)` may result in different behaviour when run in Nextflow (previously error-states were returning non-error statuses). I don't know if there's going to be consequences for this change.

kbessonov1984 replied:
If there is stop program condition after Exception is generated, then I would rather use `raise Exception(exception)` code instead of `sys.exit(1)` calls. This way by using `raise` you can "catch" the exception

kbessonov1984 replied:
I think despite all possible complexities changing to `raise Exception()` format might be better in the long run as you will know the type of exception and better capture the reason of failure

emarinier replied:
I think there might be a misunderstanding. For example, if I'm running GAS mcluster from the command line, the execution flow is like this:

`gas mcluster -t 0 -i tests/data/matrix/basic.tsv -o gas
_out --force --method single`

We first run:

```python
if __name__ == '__main__':
    run()
```

`run()` parses command line arguments with `argparse` and then runs the actual program:

```python
def run():

    cmd_args = parse_args()

    try:
        mcluster(vars(cmd_args))

    except Exception as exception:
        print("Exception: " + str(exception))
        sys.exit(1)
```

If an exception happens in `mcluster` (or something further down from there), it comes back to `run()`. There's nothing else that called it. It's the way of handling the exception when it's called from the command line. If you're calling `mcluster(...)` directly, like in the tests:

```python
    with pytest.raises(Exception) as exception:
        mcluster(args)

    assert exception.type == Exception
    assert "data/matrix/basic_MISSING.tsv does not exist or is empty" in str(exception.value)
```

then the exception handling is exactly how you're describing. It'll just raise the exception and crash like you'd expect it to.

The `try-except` block in `run()` is only used for when the program is called directly from the command line, and it's meant to preserve the functionality of error messages prior to this testing PR. Previously, the error messages were like this:

https://github.com/phac-nml/genomic_address_service/blob/0d52d98146dba3b72bde4145f40639844d497dfc/genomic_address_service/mcluster.py#L61-L67

apetkau replied:
Thanks for the clarification Eric.

>I'll also note that the change from sys.exit() to sys.exit(1) may result in different behaviour when run in Nextflow (previously error-states were returning non-error statuses). I don't know if there's going to be consequences for this change.

I think exiting with non-zero exit code is what we want, even if there's some different behaviour in Nextflow. We can address that when updating GAS in pipelines. Though I don't think there will be, as I don't think any of these exceptions ever come up in our pipeline. But if they do, I would expect them to trigger an error in Nextflow (which, with 0 exit code now, wasn't happening).

emarinier commented on genomic_address_service/classes/assign.py:
```diff
@@ -198,10 +227,10 @@ def assign(self, n_records=1000,delim="\t"):
                         alen = len(ref_address)
                         for i in range(0,len(ref_address)):
-                            addr = ".".join(ref_address[0:alen-i])
+                            addr = self.delimiter.join(ref_address[0:alen-i])
```
My best guess is that it's "address length" and probably avoids the name conflict with Python's built-in `len()` function. I'm hesitant to make the change though because I'm not very familiar with the assign function.

apetkau replied:
Thanks. I was just curious. Leaving everything as-is is okay.

emarinier commented on genomic_address_service/classes/reader.py:
```diff
@@ -19,33 +20,15 @@ def __init__(self, f, min_dist=None, max_dist=None, n_records=1000,delim="\t") -
-            header = list(df.columns)
-            num_rows = pf.num_row_groups
-
-        if len(header) == 3 and num_rows != 3:
```
There is not need to support square matrices, but also this code would cause a bug if your input was like this:

```
id  address  level_1  level_2  level_3
A  1.1.1  1  1  1
B  1.1.2  1  1  2
```

```
query_id  ref_id  dist
C  C  0
C  A  1
C  B  0
```

This (ignoring header, which it was) 3x3 pairwise distance TSV file was hitting this hard-coded 3x3 conditional and being interpreted as a matrix, which causes the IDs to be converted to floats, which crashes because they were strings. It was bugged code that only supported 3x3 matrices and did so incorrectly.

emarinier replied:
> I believe this removal is there is no need to support a square distance matrix. But, that still seems to be supported in `mcluster`. Is that true?

The file reading for `gas mcluster` is done in a completely different function:
`genomic_address_service.classes.multi_level_clustering.read_distance_matrix()`

kbessonov1984 replied:
In other words the `genomic_address_service.classes.multi_level_clustering.read_distance_matrix()` will handle all potential inputs and can differentiate between square matrix, upper or lower triangle distance matrices? For example some tools like `mash triangle` generate lower triangle distance matrix while more commonly an upper triangle matrix are generated

emarinier replied:
No, it will now only handle reading distance matrices for `gas mcluster` (not `gas call`).

Previously, if you gave `gas call` a parquet-formatted file or a distance matrix, then it would error. My understanding is that these implementations were never completed because of time constraints, so there was code that appeared to handle these data types, but did not. I have removed that code to prevent crashes and unexpected behaviour.

The `read_distance_matrix()` matrix only handles reading distance matrix files for `gas mcluster` (not `gas call`). It has nothing to do with how `gas call` handles inputs.

apetkau replied:
Thanks so much for the explanation Eric 😄

emarinier commented on genomic_address_service/classes/reader.py:
```diff
@@ -98,32 +81,13 @@ def read_matrix(self):
+        if chunk is None:
+            chunk = self.dists
 
+        return chunk
```
Updated in 25f1830

emarinier commented on genomic_address_service/classes/reader.py:
```diff
@@ -98,32 +81,13 @@ def read_matrix(self):
+        self.file_handle = open(self.fpath,'r')
+        self.header = next(self.file_handle).split(self.delim)
 
+        for chunk in self.read_pd():
```
I haven't looked much into the logic here, other than cleaning up the different file-type handling to be just text. I'll create an issue though.

kbessonov1984 replied:
Yes pandas has `chunks` option so one can read data frame by n-rows at a time, get it processed and then read another chunk. Or one can read dataframe in chunks and concat into a single dataframe. This functionality is already  but since we do not have this functionality implemented from pandas stock functions in the code, maybe let's leave current text file chunk code to support this functionality before we switch to pandas?

emarinier replied:
#41

apetkau replied:
Thanks so much for making that issue Eric 😄

emarinier commented on genomic_address_service/mcluster.py:
```diff
@@ -91,6 +103,16 @@ def run():
+
+    except Exception as exception:
+        print("Exception: " + str(exception))
+        sys.exit()
```
Updated in afdc617

kbessonov1984 replied:
Why not use `raise Exception()` instead of `sys.exit()`. This way you do not need to even `import sys` making code a bit faster.

emarinier replied:
I use `raise Exception()` all over the place:

https://github.com/phac-nml/genomic_address_service/blob/baa4d4accaf4c89f05b84443ff15f74b234cdc48/genomic_address_service/mcluster.py#L54-L68

This is just trying to preserve the behaviour of the program as it was before, as this is a PR for testing and bugfixing, not so much changing program behaviour.

`import sys` will take microseconds, which is very small for a program that we expect to take 12+ hours.

apetkau replied:
Thanks Eric 😄

emarinier commented on tests/test_GAS_utils.py:
```diff
@@ -54,23 +54,18 @@ def test_write_threshold_map():
-@pytest.mark.parametrize("outfmt", ["text", "parquet"])
-def test_write_cluster_assignments(outfmt):
+def test_write_cluster_assignments():
     memberships = {'1': 'A.B.C', '2': 'D.E.F'}
```
I'll have to think on this one. I didn't write this test, but I know the labels (conceptually) don't mean anything other than they're a unique label. There is that other specification with non-decreasing thresholds, but I'm not sure if this test was written to support labels that could be anything.

kbessonov1984 replied:
Yes the idea was to use characters to name clusters, but by default integers should be supported. I think one can use even alpha numerical characters such as plasmid clusters to do some clustering such as `AA123`. These are all edge cases and maybe skipped for now.

apetkau replied:
My reasoning for changing test to numbers is this requirement:

> Test requirement that numbers on right side of address can never be smaller than left

Using characters in the address makes that confusing. In theory you could define a way to keep the above property and have characters + numbers as part of the address, but I suspect that isn't implemented. Hence, changing this test to better match expectations of how addresses are structured.

emarinier replied:
Updated in 54635b6

kbessonov1984 commented on README.md:
````diff
@@ -66,13 +63,13 @@ This module performs assignment of new samples into an existing clustering which
 #Generate pairwise distance file
 ```
-  profile_dists -q ./example/call/query.profile.txt -r ./example/call/ref.profile.txt --outfmt pairwise --distm hamming -o ./example/call/hamming
+  profile_dists -q ./example/call/query.profile.txt -r ./example/call/ref.profile.txt --distm hamming -o ./example/call/hamming
````
I like agree that removal of redundant `--outfmt pairwise` is justified as anyways the output is a matrix which is displays pairwise distances anyways.

emarinier replied:
Wait.. maybe I shouldn't have changed this since  it's a `profile_dists` command and not a GAS command... Haha.

emarinier replied:
Reverted here (as I did not change `profile_dists`): 2a204c2

kbessonov1984 commented on genomic_address_service/call.py:
```diff
@@ -104,16 +107,21 @@ def run_call(config):
-        sys.exit()
+        exception_message = f'something went wrong with cluster assignment. Check error messages:'
+
+        for error_message in obj.error_msgs:
```
I would also add to the message what were the inputs the assign() so user would know what was the distance file and other parameters to better troubleshoot the issue. This would make error message more specific rather than being abstract

emarinier replied:
The error messages that are contained in the `err_msgs` list are more informative:

```
Exception: something went wrong with cluster assignment. Check error messages:
Error: delimiter was not found for samples ['A'].
Error: genomic address too short for samples ['B', 'C'] based on {0: 5.0, 1: 3.0, 2: 0.0}.
Error: address could not be converted to an integer for samples ['D'].
```

emarinier replied:
Here's what it looks like now:

```
Exception: something went wrong with cluster assignment
distance file: tests/data/pairwise_distances/basic.tsv
membership file: tests/data/clusters/address_errors.tsv
threshold map: {0: 5.0, 1: 3.0, 2: 0.0}
linkage method: average
delimiter: .

Check error messages:
Error: delimiter was not found for samples ['A'].
Error: genomic address too short for samples ['B', 'C'] based on {0: 5.0, 1: 3.0, 2: 0.0}.
Error: address could not be converted to an integer for samples ['D'].
cat: test_call/results.text: No such file or directory
```

Updated in de519cc

emarinier replied:
Although I suspect when this is run in IRIDA Next, the distance and membership file paths are going to be inscrutable?

kbessonov1984 replied:
I think there will be also use cases when we will run GAS in offline mode for debug and development purposes and that is where log messages with input file names could be useful. Maybe IRIDA Next team can come up with solution on using real file names instead of alphanumeric random names?

I would add to the second error message (`genomic address too short for samples ['B', 'C']`)  `based on threshold map` or `selected thresholds` as the `{0: 5.0, 1: 3.0, 2: 0.0}` dictionary might not be clear what it refers to.

emarinier replied:
It was further updated to look like this now:

```
Exception: something went wrong with cluster assignment
distance file: tests/data/pairwise_distances/basic.tsv
membership file: tests/data/clusters/address_errors.tsv
threshold map: {0: 5.0, 1: 3.0, 2: 0.0}
linkage method: average
delimiter: .

Check error messages:
Error: delimiter was not found for samples ['A'].
Error: genomic address length is incorrect for samples ['B', 'C', 'E']; expected length (3) based on thresholds {0: 5.0, 1: 3.0, 2: 0.0}.
Error: address could not be converted to an integer for samples ['D'].
cat: test_call/results.text: No such file or directory
```

apetkau replied:
> Although I suspect when this is run in IRIDA Next, the distance and membership file paths are going to be inscrutable?

When this is run through IRIDA Next the log file for GAS won't be available anyways other than for pipeline developers behind-the-scenes in Azure to help debug issues. So paths in this file referring to local paths in Azure that are only temporary is okay.

kbessonov1984 commented on genomic_address_service/classes/assign.py:
```diff
@@ -1,15 +1,20 @@
 class assign:
-    avail_methods = ["average", "complete", "single"]
+    ERROR_MISSING_DELIMITER = "delimiter was not found"
+    ERROR_TOO_SHORT = "genomic address too short"
```
Provide info what is the minimal length of the genomic address (i.e. shorter than X characters)

emarinier replied:
Updated in 0e22198

```
Error: genomic address length is incorrect for samples ['B', 'C', 'E']; expected length (3) based on thresholds {0: 5.0, 1: 3.0, 2: 0.0}.
```

```
Exception: something went wrong with cluster assignment
distance file: tests/data/pairwise_distances/basic.tsv
membership file: tests/data/clusters/address_errors.tsv
threshold map: {0: 5.0, 1: 3.0, 2: 0.0}
linkage method: average
delimiter: .

Check error messages:
Error: delimiter was not found for samples ['A'].
Error: genomic address length is incorrect for samples ['B', 'C', 'E']; expected length (3) based on thresholds {0: 5.0, 1: 3.0, 2: 0.0}.
Error: address could not be converted to an integer for samples ['D'].
cat: test_call/results.text: No such file or directory
```

It's also now more general for addresses that are too long as well.

kbessonov1984 replied:
Excellent, I like this detailed message. Thank you for the update

kbessonov1984 commented on genomic_address_service/classes/assign.py:
```diff
@@ -153,31 +184,29 @@ def get_threshold_idx(self,dist):
-                columns=None,
-                storage_options=None,)
+        else:
+            message = f'{f} does not have a valid extension'
```
Add info what extension was provided by the user and print out what are valid extensions

emarinier replied:
I agree that this is confusing. The code was already doing what you were suggest, but in the `guess_file_type` part. I've cleaned up the code to this instead:

```python
def check_file_type(self,f):
    extension = os.path.splitext(f)[1]
    valid_extensions = list(EXTENSIONS.keys())

    if not extension in valid_extensions:
        message = f'{f} does not have a valid extension {valid_extensions}'
        raise Exception(message)

    return EXTENSIONS[extension]

def read_data(self, f):
    self.check_file_type(f)
    df = pd.read_csv(f, header=0, sep="\t", low_memory=False)

    return df
````

updated (and more changes) here: 69849c9

kbessonov1984 commented on genomic_address_service/classes/assign.py:
```diff
@@ -153,31 +184,29 @@ def get_threshold_idx(self,dist):
         file_type = self.guess_file_type(f)
-        if file_type == 'text':
+
+        if file_type == TEXT:
```
You should also test if the TEXT file has delimiter other than `\t` tab delimiter. In that case pandas will not be able to identify columns so we can then fail the input saying that expected delimiter was not found. Are we supporting also csv files?

emarinier replied:
See here: https://github.com/phac-nml/genomic_address_service/pull/38#issuecomment-2790493569

kbessonov1984 commented on genomic_address_service/classes/reader.py:
```diff
@@ -98,32 +81,14 @@ def read_matrix(self):
-            self.file_handle = open(self.fpath,'r')
-            self.header = next(self.file_handle).split(self.delim)
-        elif ftype == 'parquet':
-            self.file_handle = ParquetFile(self.fpath)
```
Why parquet file type was removed? What was the logic behind it?

emarinier replied:
It wasn't working. If you provided a validly formatted parquet file, GAS call wouldn't process it.

Here's the old code block:

https://github.com/phac-nml/genomic_address_service/blob/0d52d98146dba3b72bde4145f40639844d497dfc/genomic_address_service/classes/reader.py#L100-L123

In the second conditional block there's 2 cases handled:

`if ftype == 'text' and dist_type == 'pd':` - A pairwise distances text file.
`elif ftype == 'text' and dist_type == 'matrix':` - A matrix distances text file (which did not work for other reasons).

There was no case for something like:

`if ftype == 'parquet':`

So the code opened up Parquet files but then didn't do anything with them. I'm told it was left unfinished because of time constraints.

emarinier commented:
The comment about checking for tab files was deleted, but here's the response and work:

I don't think there's a simple and fast way to check the entire file. It's also difficult because of the three TSV-formatted inputs, distance matrices (`gas mcluster`) and cluster files (`gas call`) have a variable number of columns. Pairwise distance columns (`gas call`) have a fixed number of columns. There's also no unambiguous way of writing a function that tells you what the delimiter is, because one can write gibberish that'll create ambiguous data.

I wrote some code to check the headers for the minimum number of tabs. This will catch some errors, but not all of them.

```
gas mcluster -t 5,3,0 -i tests/data/matrix/csv.text -o gas_out --force --method single
Exception: tests/data/matrix/csv.text does not appear to be a properly TSV-formatted file
```

```
gas call -t 5,3,0 -o test_call -r tests/data/clusters/csv.text -d tests/data/pairwise_distances/basic.tsv --force --delimiter "."
Exception: tests/data/clusters/csv.text does not appear to be a properly TSV-formatted file
```

```
gas call -t 5,3,0 -o test_call -r tests/data/clusters/small.tsv -d tests/data/pairwise_distances/csv.text --force --delimiter "."
Exception: tests/data/pairwise_distances/csv.text does not appear to be a properly TSV-formatted file
```

Updated here: 09d9212
Small fix for pairwise here: 17b79a5

emarinier commented:
> These tests check for `results.text` and `clusters.text`. But, we had changed the extensions to TSV for these so I'm not sure how the tests are still working. Could you check into this?

We didn't make the change in this PR. It was added to another issue: #40

emarinier commented on CHANGELOG.md:
```diff
@@ -3,6 +3,24 @@
+## [0.1.6] - 2025-04-09
+
+### Added
+- Numerous integration tests.
```
Updated in 112f9db

emarinier commented on CHANGELOG.md:
```diff
@@ -3,6 +3,24 @@
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
+## [0.1.6] - 2025-04-09
```
Updated in 6b9f5d1

emarinier commented:
> ### 2. Tests expected to fail
> 
> There are two tests that are marked as expected to fail:
> 
> https://github.com/phac-nml/genomic_address_service/blob/6ff1b9931f994b17d3866cbe0a2f5abf2db51cba/tests/test_reader.py#L90-L96
> 
> https://github.com/phac-nml/genomic_address_service/blob/6ff1b9931f994b17d3866cbe0a2f5abf2db51cba/tests/test_reader.py#L125-L136
> 
> Could you look at either fixing these or removing the tests if they are not needed?

Updated in bb84fda

- The first test (Parquet files): this is no longer need because we don't support Parquet files.
- The second test (file not read properly): I think this test might've previously tried testing square distance matrix input to GAS call (which is what the reader function here is used for), but the code for that wasn't working correctly. Since we no longer support that kind of input, I've removed the test.

emarinier commented:
> ### 1. End-to-end tests with pytest-workflow
> 
> There are some end-to-end tests that run `gas call` and `gas mcluster` at https://github.com/phac-nml/genomic_address_service/blob/main/tests/test_workflows.yml
> 
> This uses https://pytest-workflow.readthedocs.io/en/stable/ (you install `pip install pytest-workflow` and then running `pytest` will run the commands in the above YAML file).

I've double-checked that these workflow tests still pass (in part because outputs will be changed later for #40).

apetkau reviewed (approved):
This is amazing work. Thanks so much Eric 😄
